### ISSUE TITLE
Expand float_cmp to structs with PartialEq

### ISSUE THREAD

**Issue by RazrFalcon** [Manishearth/rust-clippy#1685]
Example:
```rust
#[derive(PartialEq)]
struct Data {
    v: f64,
}

fn main() {
    let d1 = Data { v: 1.0 };
    let d2 = Data { v: 2.0 };

    // here should be a warning
    println!("{}", d1 == d2);
}
```

**Comment by clarcharr** [comment]
But this should only be a warning if the structure also implements `PartialOrd`.

**Comment by RazrFalcon** [comment]
Doesn't `PartialEq` implemented without comparing the floats?

**Comment by clarcharr** [comment]
@RazrFalcon generally, the suggestion for floats is that instead of doing something like `x == y`, you'd do `(x - y).abs() < z`.

Although in this case, we have no idea if that's necessarily possible with these sorts of structs. You could warn on `PartialEq` for structs that are just floats, though, suggesting other alternatives.

**Comment by oli-obk** [comment]
I think this should not lint on the comparison site, but on the struct definition. Any struct deriving `PartialEq` and having float fields should be linted against.

**Comment by cschwan** [comment]
> I think this should not lint on the comparison site, but on the struct definition. Any struct deriving `PartialEq` and having float fields should be linted against.

I think it's a bit more complicated than this, for instance consider comparing two `Vec<f64>`; in this case I would like the linter to complain at the comparison site, but I agree that when a struct has non-generic float fields it's best linted on the struct definition.
